### ISSUE TITLE
Unification fails in the presence of types depending on evars

### ISSUE THREAD

**Issue by RalfJung** [coq/coq#7597]
#### Version

8.8.0

#### Operating system
Linux

#### Description of the problem
The following fails to compile:
```
Fixpoint tfun (x : nat) : Type :=
  match x with
  | O => nat
  | S x => nat -> tfun x
  end.

Definition foo (x : nat) (f : tfun x) : nat := O.

Goal exists a b, foo 1 (fun x => x) = foo a b.
Proof. eexists. eexists. reflexivity. Qed.
```
saying
```
Error: Unable to unify "?b" with "fun x : nat => x"
(unable to find a well-typed instantiation for "?b": cannot ensure that
"nat -> nat" is a subtype of "tfun ?a").
```

The unification problem in the last line is `foo 1 (fun x : nat => x) = foo ?a ?b`, which should not be hard to solve. However, Coq tries to unify `?b` first, which of course doesn't work -- only after unification of `?a`, we know enough about even the type of `?b` to go on.

Cc @herbelin

**Comment by gares** [comment]
This problem goes outside what unification can do. To find a solution you would need to say `?a := S ?a1`for some `?a1`. This is a class of problems similar to what Canonical Structure can solve, but given that there is no record/projection here you can't declare a CS. It falls in the class of problems you may solve with Unification Hints (you find a paper of mine if you google it, not implemented in Coq, will eventually be the unification algorithm I'm working on in elpi).

It is not a wontfix, it is a "I'd love to fix". But you can't really hold your breath for this one.

**Comment by herbelin** [comment]
@gares: I think that a possible approach to this particular problem would be not to claim that it is a failure in the flex-rigid case in the presence of unresolved `nat -> nat == match ?a with ... end` but instead to add the problem as a constraint (line `ise_try evd [consume_stack on_left apprF apprR; eta]`). It is not clear that doing so will not break elsewhere due to another weakness of the algorithm (I remember that I tried variants of such postponenent which was revealing other problems), but we have to start by some end of the current limitations, otherwise we shall never make progresses.

Another approach, which probably amounts to the kind of unification hints you suggest to add, would be to try both `0` and `S` for `?a` and to see that only `S` works.

**Comment by JasonGross** [comment]
I think this is a specific example of a class of problems I think unification should succeed on: if unification would succeed if I first filled in a maximal subset of all evar instantiations which result from considering the pattern fragment (maximal under the constraint that the simultaneous instantiation should be well-typed without inducing additional constraints), then unification should succeed.

Said another way, order of arguments should not matter when the entire problem can be solved just by considering the pattern fragment.

**Comment by gares** [comment]
I'm sorry I looked at the unification problem too quickly. It seems it is just that the arguments of the two foo are unified in the wrong order, from right to left. Ie ?b is unified before ?a is...

**Comment by RalfJung** [comment]
> To find a solution you would need to say ?a := S ?a1for some ?a1

I must be missing something, because I don't see how that's a requirement. From what I see. unifying `foo 1 (fun x : nat => x)` and `foo ?a ?b` at type `nat` could proceed something like
* Compare the heads: They are the same term at the same type! Both sides have `foo` at type `forall x : nat, tfun x -> nat`.
* Now iterate over the arguments, left-to-right, and unify them
  * First sub-unification problem: Unify `1` and `?a` at type `nat`. That should be easy.
  * Second and last sub-unification problem: unify `(fun x : nat => x)` and `?b` at types `tfun 1` and `tfun ?a`, respectively. Well, `?a` is already picked to be `1`, so just expanding the evars should reveal that the type is the same on both sides (namely, `tfun 1`). Now, unifying `(fun x : nat => x)` and `?b`  should be trivial.

Admittedly, I don't know much about unification. But I cannot understand how this can be so hard. It seems that Coq tries to unify right-to-left, and of course since the later arguments can depend on the earlier ones, that leads to a much harder problem.

---

Interestingly, using `exact eq_refl` instead of reflexivity seems to solve the problem here:
```
Fixpoint tfun (x : nat) : Type :=
  match x with
  | O => nat
  | S x => nat -> tfun x
  end.

Definition foo (x : nat) (f : tfun x) : nat := O.

Goal exists a b, foo 1 (fun x => x) = foo a b.
Proof. eexists. eexists. exact eq_refl. Qed.
```
So it seems like Coq actually *can* solve this, under the right circumstances?

**Comment by gares** [comment]
This means that the unification algorithm (evarconv.ml) does unify in the right order (first `?a:=1`, then ?b:=(fun x=>x) and this assignment is well types since the type of `?b` is now `tfun 1` and not `tfun ?a` as in the error message, and `tfun 1 = nat -> nat` ).

On the contrary reflexivity does it in the wrong order. I've no clue why. Maybe it uses the old unification algorithm (unification.ml)? @herbelin does the old algorithm go right to left?

**Comment by SkySkimmer** [comment]
reflexivity is old unification..
You can see that `apply eq_refl.` also fails.

**Comment by mattam82** [comment]
Indeed, that’s one of the quirks of the old unification, which does not substitute eagerly the solutions of previous problems, hence fails in many such dependent cases.

**Comment by RalfJung** [comment]
> You can see that apply eq_refl. also fails.

Yeah, and in my real code it's `simple apply` (from typeclass resolution) that's the culprit.

Maybe I can hack it with a `Hint Extern`...

**Comment by gares** [comment]
> Indeed, that’s one of the quirks of the old unification, which does not substitute eagerly the solutions of previous problems, hence fails in many such dependent cases.

This is isomorphic to "it is totally broken". Please, stop the suffering and replace it.

**Comment by herbelin** [comment]
> This is isomorphic to "it is totally broken". 

My vague memory is that the motivation of Chet M. for postponing substitutions was to make `auto` faster.

**Comment by maximedenes** [comment]
When a change is incorrect, I don't think it should qualify as "faster", otherwise I can easily submit a patch that will make Coq a lot faster :)

**Comment by herbelin** [comment]
> When a change is incorrect

Why do you say it is incorrect? Postponing substitutions does not mean that substitutions are not done.

> otherwise I can easily submit a patch that will make Coq a lot faster :)

Users of `auto` shall be happy if you submit such a patch.

**Comment by gares** [comment]
`Ltac auto := fail. (* very fast *)`

**Comment by herbelin** [comment]
I should add that the motivation for my comment was just to replace the choices made by Chet more than 20 years ago in its context. I don't think that anyone make deliberatedly wrong choices. Choices are generally motivated and at least good from the point of view of the person who defend them. Then, we can discuss how choices are good from a more objective/collective/consensual/universal point of view, but again, I don't think that it would be fair to blame choices made at some time of the history without understanding the original motivations for these choices.

**Comment by mattam82** [comment]
I think it simply reveals that at the time there was no use case of such
dependent unifications at the tactic level. And yes today it looks like a
counter-productive choice. I don’t see how it could be more efficient to
not do the eager expansion and wait for later (you only work with what
appears to be a less constrained problem) but I would guess at the time
there was no evar_map, and the metamap probably did not have a
normalization function to get the up-to-date type of a meta. All this was
and is still done in w_merge after unification proper.
Le ven. 25 mai 2018 à 19:50, Hugo Herbelin <notifications@github.com> a
écrit :

**Comment by maximedenes** [comment]
@herbelin I agree with you, I didn't mean to comment on the change done at the time, on a codebase that was very different. I was rather trying to make clear the terminology we use to describe the current status, sorry if it was ambiguous.

**Comment by herbelin** [comment]
Sorry then, because I did not understand that you were trying to clarify a terminology (then presumably either of the word "fast" or of the word "broken"?). I saw gratuitous irony denying the contribution of Chet and I did not understand the justification of it. If I remember correctly, Chet wrote both unification algorithms and if he had wanted to use immediate substitution in the tactic unification algorithm, he could have done it as he had done it in the type inference unification algorithm.

I know by experience how difficult it is to express his/her mind while anticipating possible ambiguities or misunderstandings. I wish we eventually understand better each other. You have good intuitions in general and that's a pity that you often show to my eyes only the superficial if not most controversial parts of them.

**Comment by herbelin** [comment]
Sorry, this is entirely a digression/hijacking and it can eventually be hidden, but it happens that a friend is just reading a book entitled [La societé comme verdict](https://www.fayard.fr/la-societe-comme-verdict-9782213655833):

The back cover says: « La société assigne des places. Elle énonce des verdicts, qui s’emparent de nous et marquent nos vies à tout jamais. Elle installe des frontières et hiérarchise les individus et les groupes. 
La tâche de la pensée est de porter au jour ces mécanismes d’infériorisation et la logique de la domination et de la reproduction sociales. C’est à un véritable renouvellement de l’analyse des classes, des trajectoires, des identités et du rôle central et ambivalent des institutions (notamment le système scolaire, le droit, la politique...) dans leur fabrication que nous convie Didier Eribon. Avec pour horizon l’idée que seule une démarche qui place au centre de ses préoccupations le problème des déterminismes par lesquels nos vies sont régies peut nous permettre d’ouvrir la voie à une politique de l’émancipation. »

I wish that our world gets more and more aware of these determinisms, so that a kind of human relations, freed from conformism to social roles or rules, is possible.
